### ISSUE TITLE
Port protocols don't include udp when using ss on ubuntu 14.04

### ISSUE THREAD

**Issue by Afterglow** [inspec/inspec#2241]
### Description

Checking protocols of ports does not appear to correctly detect ports listening on UDP

### InSpec and Platform Version

Inspec Version: `1.41`
Running within a `bento/14.04` VM

### Replication Case

1) Using Vagrant...

```
Vagrant.configure("2") do |config|
  config.vm.box = "bento/ubuntu-14.04"
end
```

2) Install inspec 1.41
`sudo apt-get update`
`sudo apt-get install curl`
`curl https://omnitruck.chef.io/install.sh | sudo bash -s -- -P inspec -v 1.41`

3) Running `netstat -tulpen` per the way the `ports_via_netstat` method works should show that port `111`(rpcbind) is listening on TCP and UDP.

4) Write a simple inspec test

```ruby
describe port(111) do
  its('protocols') { should include 'udp' }
  its('protocols') { should include 'tcp' }
end
```

5) Run the test with inspec 1.41

```
vagrant@vagrant:~$ sudo inspec exec ./test.rb
WARN: Unresolved specs during Gem::Specification.reset:
      rake (>= 0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.

Profile: tests from ./test.rb
Version: (not specified)
Target:  local://


  Port 111
     ∅  protocols should include "udp"
     expected ["tcp"] to include "udp"
     ✔  protocols should include "tcp"

Test Summary: 1 successful, 1 failure, 0 skipped
```

6) Move `ss` out of the path `sudo mv /bin/ss ~/` and run the tests again

```
vagrant@vagrant:~$ sudo inspec exec ./test.rb
WARN: Unresolved specs during Gem::Specification.reset:
      rake (>= 0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.

Profile: tests from ./test.rb
Version: (not specified)
Target:  local://


  Port 111
     ✔  protocols should include "udp"
     ✔  protocols should include "tcp"

Test Summary: 2 successful, 0 failures, 0 skipped
```

**Comment by Afterglow** [comment]
is further information required for this?

**Comment by adamleff** [comment]
No additional information is needed, @Afterglow - we simply haven't had time to address this issue. If you're willing to take a stab at submitting a PR, we'd be happy to review.

**Comment by adamleff** [comment]
Hi, @afterglow! I had some time to look at this today.

Unfortunately, this appears to be an issue with Ubuntu's iproute2 package for 14.04. Since `netstat` is officially deprecated, InSpec looks for `ss` first, and if it's there, we use it. Here's the output of `ss -tulpen` on a 14.04 instance:

```
root@vagrant:~# ss -tulpen | grep 111
tcp    UNCONN     0      0                      *:111                   *:*      users:(("rpcbind",563,6)) ino:9119 sk:ffff88003bca4000 <->
tcp    UNCONN     0      0                     :::111                  :::*      users:(("rpcbind",563,9)) ino:9124 sk:ffff88003bca8000 <->
tcp    LISTEN     0      128                    *:111                   *:*      users:(("rpcbind",563,8)) ino:9123 sk:ffff88003bcc4000 <->
tcp    LISTEN     0      128                   :::111                  :::*      users:(("rpcbind",563,11)) ino:9126 sk:ffff88003c260000 <->
```

... even though two of those entries are actually UDP, all four are labeled TCP. If you drop the `t` and run `ss -ulpen` you'll see the rpcbind ports listed there.

Compare this with the output of `ss -tulpen` on a Ubuntu 16.04 host:

```
root@vagrant:~# ss -tulpen | grep 111
udp    UNCONN     0      0         *:111                   *:*                   users:(("rpcbind",pid=1903,fd=6)) ino:20360 sk:5 <->
udp    UNCONN     0      0        :::111                  :::*                   users:(("rpcbind",pid=1903,fd=9)) ino:20363 sk:7 v6only:1 <->
tcp    LISTEN     0      128       *:111                   *:*                   users:(("rpcbind",pid=1903,fd=8)) ino:20362 sk:8 <->
tcp    LISTEN     0      128      :::111                  :::*                   users:(("rpcbind",pid=1903,fd=11)) ino:20365 sk:9 v6only:1 <->
```

... you can see that the protocol is listed correctly. And InSpec is properly parsing it:

```
inspec> port(111).entries
=> [#<struct  port=111, address="0.0.0.0", protocol="udp", process="rpcbind", pid=1667, listening?=nil>,
 #<struct  port=111, address="::", protocol="udp6", process="rpcbind", pid=1667, listening?=nil>,
 #<struct  port=111, address="0.0.0.0", protocol="tcp", process="rpcbind", pid=1667, listening?=nil>,
 #<struct  port=111, address="::", protocol="tcp6", process="rpcbind", pid=1667, listening?=nil>]
inspec>
inspec>
inspec> port(111).where { protocol == 'udp' }.entries
=> [#<struct  port=111, address="0.0.0.0", protocol="udp", process="rpcbind", pid=1667, listening?=nil>]
```

I'm not honestly sure the best way to proceed. I don't think we want to modify InSpec to work around a pretty clear bug in an OS package, but I don't believe there's an updated iproute2 package available for 14.04. I don't know when `ss` was fixed to properly report udp ports as udp, so it would be difficult to fallback to netstat based on the iproute2 package version. I suppose we could force a fallback to `netstat` if on ubuntu < 16. @chris-rock and @arlimus: do you have any thoughts on this?

**Comment by miah** [comment]
__What’s happening? Why was this issue closed?__
This issue was closed due to some much needed review of legacy issues or issues that were spawned in older versions of InSpec, i.e. < v3.

__Why do I care?__
You would care about this if this was an issue that you are still seeing and/or feel needs to be addressed in the current version of InSpec.

__What do I need to do?__
If this issue is no longer important, no further action is necessary. However, if you think this is something that should be addressed, please open a new issue and refer to the original issue in the description.
